src/coreclr/src/vm/callingconvention.h

```diff
@@ -1550,6 +1550,29 @@ void ArgIteratorTemplate<ARGITERATOR_BASE>::ComputeReturnFlags()
             }
 #endif
 
+#if defined(_TARGET_AMD64_) && !defined(UNIX_AMD64_ABI)
+            int vectorSize = thValueType.AsMethodTable()->GetVectorSize();
+            if (vectorSize == 16) /* Check if it is a Vector128<T> or 128 bit Vector<T>. */
+            {
+                // If FEATURE_SIMD is disabled, this is a normal 16 byte structure
+                EEJitManager *jitMgr = ExecutionManager::GetEEJitManager();
+                if (jitMgr->LoadJIT())
+                {
+                    CORJIT_FLAGS cpuCompileFlags = jitMgr->GetCPUCompileFlags();
+                    if (cpuCompileFlags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_FEATURE_SIMD))
```

Comment:
It does not feel right to vary calling convention based on these flags. Would it be better for  `CORJIT_FLAG_FEATURE_SIMD` to only control whether hardware intrinsics are enable, but not alter calling convention?

Reply:
IIRC, we don't have support for any of the `TYP_SIMD` stuff without this flag. So there might be a bit more involved in recognizing the type and getting things to work in that scenario.

Reply:
Should this flag be unconditionally set on Windows x64 then?

Reply:
I believe that is currently the case for the mainline build: https://github.com/dotnet/runtime/blob/5040afeb94855b630ba822a9d62162c3652fe392/src/coreclr/src/jit/CMakeLists.txt#L13-L16

@CarolEidt might know if there are any cases where it is or could be disabled (at least from a quick grep, it is disabled for `armelnonjit`, `linuxnonjit`, and `protononjit`).

Reply:
I added this weird quirk for the no simd testing scenario, but as you say, its not really supported or generally tested. Would there be objection to removing the ability to turn off the ability to disable feature simd on Windows X64?

Reply:
> Would there be objection to removing the ability to turn off the ability to disable feature simd on Windows X64?

What about the other platforms (i.e. ARM64)? Is there any supported calling convention for x86 where `FEATURE_SIMD` would be required (whether Windows, Linux, or MacOS)?
* At least from an initial search, ARM64 ABI looks to use FP/SIMD registers and might fit in a similar bucket. But the x86 ABI (outside the currently unsupported `__vectorcall`) was defined before SSE and so will never pass in a SIMD register.

Reply:
It is unlikely we will ever support an X86 platform where we don't have basic SSE support, as such I think disabling SIMD support doesn't make much sense except for esoteric test scenarios that it turns out we don't use much anyways. A much more interesting question is for ARM64 on Linux and ARM32.

Arm32 doesn't support simd today, and thus an abi shift based on FEATURE_SIMD support will be necessary for when we add support for `Vector<T>`, as the abi of `Vector<T>` is fundamentally based on using a vector register. However, I think it is reasonable for the Vector<T> abi to vary based on weird switches, as it really is supposed to be dynamic.

For Arm64, we support AdvSimd and floating point in the baseline, and according to the comment in pal\src\misc\jitsupport.cpp, we don't support platforms without that support, but we do actually have the checks that call into the kernel to ask if such support exists. It would be quite reasonable for a `Vector<T>` to use SVE in the future, and using SVE might imply a different calling convention. That adjusted calling convention would be problematic, as the architectural max size of a SVE register is 2048 bits, and would thus bloat our transition block structure enormously. 

As far as "supported calling convention on a platform", well, the SysV Intel abi on Linux/Mac is a bit peculiar around this. It allows for use of the registers as long as all compiled is compiled as machine aware, but the struct types work regardless. For instance, if you compile a library as Avx aware in clang on Linux, and expose an abi that manipulates an __m256, and call that function from another binary that was not compiled as Avx aware but does use the __m256 structure, then the same C++ compiler will generate conflicting abis, and things will break. Exposing interop support, will cause us to need to be able to talk to such issues, and reasons like that are why this work will not directly enable vector marshalling support.

For Windows, matters are a bit simpler. On Windows the C++ abi does not vary by compiler switch. The compiler switches will however, drive if a __m256 type is useable at all.

Arm32/64 abi requires use of SIMD registers for the SIMD types. `Vector128<T>` should only be passed in SIMD registers, etc.

Reply:
> no simd testing scenario

We should have a way to disable hardware accelerated Vector<T> and the hardware intrinsics IsSupported methods so that people can test their fallback paths. But the calling convention can stay the same, even when these are disabled.

Reply:
I agree with @jkotas here - we need to have a way to test fallback paths (though I would note that this should really only impact the ability to test those fallback paths on a platform that would otherwise not require them), and it may be that changing the behavior of `FEATURE_SIMD` in this way will require additional changes, but if so those changes should be made.

Reply:
> We should have a way to disable hardware accelerated Vector and the hardware intrinsics IsSupported methods so that people can test their fallback paths.

In the case of HWIntrinsics, this is done via the `COMPlus_Enable{ISA}=0` or `COMPlus_EnableHWIntrinsic=0` flags (see https://github.com/dotnet/runtime/blob/fcd862e06413a000f9cafa9d2f359226c60b9b42/src/coreclr/src/jit/jitconfigvalues.h#L226-L272).
I don't believe we have a flag for `Vector<T>` today.

Reply:
Correct; that's what `FEATURE_SIMD` was generally used for, and I was just indicating that we need a way to disable `Vector<T>` without impacting the calling convention, and that doing so will probably require additional changes.